NIP-5C: Scrolls (WASM programs) (#2281)
nostr-protocol/nips: A5.md

### `nostr.req_new() -> i32`

Create a new empty request handle.

### `nostr.req_add_author(req: i32, pubkey_ptr: i32) -> void`

Add a pubkey to the `authors` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_author_hex(req: i32, pubkey_hex_ptr: i32) -> void`

Same as `req_add_author` but with a hex-encoded pubkey string. Length is assumed to be 64 characters.

### `nostr.req_add_id(req: i32, id_ptr: i32) -> void`

Add an id to the `ids` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_id_hex(req: i32, id_hex_ptr: i32) -> void`

Same as `req_add_id` but with a hex-encoded id string. Length is assumed to be 64 characters.

### `nostr.req_add_kind(req: i32, kind: i32) -> void`

Add a kind integer to the `kinds` filter.

### `nostr.req_add_tag(req: i32, tag_ptr: i32, tag_len: i32, value_ptr: i32, value_len: i32) -> void`

Add a value to a tag filter. `tag_ptr/tag_len` points to the tag name (e.g. `"p"` -- will be added to the filter as `"#p"`); `value_ptr/value_len` points to the tag value to match. The value is treated as a string.

### `nostr.req_add_tag_bin32(req: i32, tag_ptr: i32, value_ptr: i32) -> void`

> Same as `req_add_tag` but the value is a 32-byte binary buffer that the host will convert to hex. `value_ptr` must point to a 32-byte buffer. This is useful for `#p`, `#e`, and other tag filters that need pubkey or event ID values.

Reviewer: this should mention that the `tag_ptr` should point to a single byte string

Reviewer: in your host runtime it ask for the tag len. `req_add_tag_bin32: (req: number, tPtr: number, tLen: number, vPtr: number): number`

Author: I'll fix this on the spec. This was me trying to change it to be a single letter instead of a pointer, but @alexgleason was going to complain later that he wanted to use multi-letter tag filters, so I went back.

> Same as `req_add_tag` but the value is a 32-byte binary buffer that the host will convert to hex. `value_ptr` must point to a 32-byte buffer. This is useful for `#p`, `#e`, and other tag filters that need pubkey or event ID values.

Reviewer: ```suggestion
Same as `req_add_tag` but the value is a pointer to a 32-byte binary buffer that the host will convert to hex. This is useful for `#p`, `#e`, and other tag filters that need pubkey or event ID values.
```

The globally exported `on_event` function will be called with the results.

- The module must call `nostr.drop(event)` on each non-zero event handle when done with it.
- If `nostr.req_close_on_eose` was set, the host drops the subscription handle after the EOSE callback. The module must not drop it.

To cancel a live subscription call `nostr.drop(sub)`.

### `nostr.event_get_id(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the event ID.

### `nostr.event_get_id_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex event ID.

### `nostr.event_get_pubkey(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the public key of the event author.

### `nostr.event_get_pubkey_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex public key of the event author.

### `nostr.event_get_kind(event_handle: i32) -> i32`

Returns the kind integer directly.

### `nostr.event_get_created_at(event_handle: i32) -> i32`

Returns the unix timestamp directly.

### `nostr.event_get_content(event_handle: i32) -> i32`

> Returns a pointer to a buffer containing the event content.

Reviewer: same as `event_get_tag_item` below, it should mention that it add the string length before it. Which is 4 bytes (32-bit unsigned integer big-endian)

### `nostr.event_get_tag_count(event_handle: i32) -> i32`

Returns the total number of tags on this event.

### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: i32) -> i32`

Returns the number of items in the tag at `tag_index`.

### `nostr.event_get_tag_item(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

> Returns a pointer to a buffer containing the item at `(tag_index, item_index)`.

Reviewer: the item is string, how would I know its length?

Reviewer: Okay, I see that you add the length before the string. This should be mentioned.

```ts
  function writeHostStr(str: string, prefixLen: boolean = true): number {
    const encoded = new TextEncoder().encode(str)
    return writeHostBuf(encoded, prefixLen)
  }

  function writeHostBuf(buf: Uint8Array, prefixLen: boolean): number {
    const prefixOffset = prefixLen ? 4 : 0
    const ptr = (instance.exports.alloc as (size: number) => number)(buf.length + prefixOffset)
    const view = new DataView(memory.buffer)
    if (prefixLen) view.setUint32(ptr, buf.length, false)
    new Uint8Array(memory.buffer, ptr + prefixOffset, buf.length).set(buf)
    return ptr
  }
```

event_get_tag_item:
```ts
(ev: number, ti: number, ii: number): number => {
    const val = get(ev, "event").tags?.[ti]?.[ii]
    console.log(
      `[nprogram] event_get_tag_item(ev=${ev}, ti=${ti}, ii=${ii}) => "${val?.slice(0, 16) ?? "null"}..."`
    )
    return val != null ? writeHostStr(val) : 0
  }
```

Reviewer: You actually mentioned this above in "String Convention". IDK, I'm not sure if you should add it here also or not.

### `nostr.event_get_tag_item_bin32(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

> Same as `event_get_tag_item`, but returns a 32-byte buffer of the item if it happened to be a pubkey or an event id; 0 otherwise.

Reviewer: ```suggestion
Same as `event_get_tag_item`, but returns a pointer to a 32-byte buffer of the item if it happened to be a pubkey or an event id; 0 otherwise.
```

### `nostr.event_get_tag_item_by_name(event_handle: i32, name_ptr: i32, name_len: i32, item_index: i32) -> i32`

Finds the first tag whose name (item 0) matches the string at `name_ptr/name_len`, then returns a pointer to a buffer containing item `item_index` from that tag; `0` if no matching tag is found.

### `nostr.event_get_tag_item_by_name_bin32(event_handle: i32, name_ptr: i32, name_len: i32, item_index: i32) -> i32`

> Same as `event_get_tag_item_by_name`, but returns a 32-byte buffer of the value if it happened to be a pubkey or an event id; 0 otherwise.

Reviewer: ```suggestion
Same as `event_get_tag_item_by_name`, but returns a pointer to a 32-byte buffer of the value if it happened to be a pubkey or an event id; 0 otherwise.
```

### `nostr.display(event: i32) -> void`

Render an event through the client's native note renderer. The event handle is not consumed.

### `nostr.log(ptr: i32, len: i32) -> void`

Emit a log message to the host's debug console or developer tooling. The string at `ptr/len` is the message.

### `nostr.drop(handle: i32) -> void`

> Releases any handle: unconsumed request, active subscription (cancels it), individual event, or list (also releases all contained event handles).

Reviewer: what does "or list (also releases all contained event handles)" means? how can I pass a list as a handle? only handles is req, event and subscription

Reviewer: The host runtime only works with handles, so I think "list" here should be removed.

```ts
  function dropHandle(h: number): void {
    if (h === 0) return
    const entry = handles.get(h)
    if (!entry) return
    if (entry.type === "sub" && entry.data.closer) {
      entry.data.closer.close()
    }
    handles.delete(h)
  }
```

Author: This is also a remainder of an old approach to event fetching. Thank you for spotting.
